vpr/src/place/cut_spreader.h

```diff
@@ -0,0 +1,73 @@
+#ifndef VPR_SRC_PLACE_LEGALIZER_H_
+#define VPR_SRC_PLACE_LEGALIZER_H_
+
+#include "vpr_context.h"
+
+class AnalyticPlacer;
+struct t_logical_block_type;
+
+// Cut-spreader, as described in HeAP/SimPL papers
+class CutSpreader {
+  public:
+    // can refactor to placement_hints data structure instead of AnalyticPlacer.
+    CutSpreader(AnalyticPlacer* analytic_placer, VprContext& vpr_ctx, t_logical_block_type_ptr cell_t);
+
+    void run();
+
+    void strict_legalize();
+
+  private:
+    int32_t dont_solve = std::numeric_limits<int32_t>::max();
+
+    AnalyticPlacer* ap;
+    const DeviceContext& device_ctx;
+    const ClusteredNetlist& clb_nlist;
+    PlacementContext& place_ctx;
+    t_logical_block_type_ptr blk_type;
+
+    struct MacroExtent {
```

Comment:
Commenting data members is particularly important.

Reply:
Explain what x0,y0, x1,y1 are.  Is one the lower-left and the other the upper-right?  Or are they unordered where x0,y0 could be either corner?  Are they relative to the placement of the macro head or absolute values (i.e. if the macro head is at (1,6) and it's a two block macro with the second block offset by y_offset = -1, what are (x0,y0) and (x1,y1)?

Reply:
could use vtr::rect for this.